components/validator/challenger.go

```diff
@@ -24,232 +31,493 @@ type ProofFetcher interface {
+			}
+
+			logs, err := c.l1Client.FilterLogs(c.ctx, query)
+			if err != nil {
+				c.log.Error("failed to get event logs related to outputs", "err", err)
+				break Loop
+			}
+
+			for _, vLog := range logs {
+				// for OutputSubmitted event
+				if !c.cfg.ChallengerDisabled && vLog.Address == c.cfg.L2OutputOracleAddr {
+					ev := NewOutputSubmittedEvent(vLog)
+					// handle output
+					c.wg.Add(1)
+					go c.handleOutput(ev.OutputIndex)
+					continue
+				}
 
-	if c.checkpoint == nil {
-		submissionIntervalSeconds := new(big.Int).Mul(c.submissionInterval, c.l2BlockTime)
-		waitingOutputNum := new(big.Int).Div(c.finalizationPeriodSeconds, submissionIntervalSeconds)
-		if latestOutputIndex.Cmp(waitingOutputNum) != 1 {
-			c.checkpoint = common.Big1
-		} else if waitingOutputNum.Cmp(common.Big0) == 0 {
-			c.checkpoint = latestOutputIndex
-			return nil, nil
-		} else {
-			c.checkpoint = new(big.Int).Sub(nextOutputIndex, waitingOutputNum)
+				// for ChallengeCreated event
+				if vLog.Address == c.cfg.ColosseumAddr {
+					// handle challenge
+					ev := NewChallengeCreatedEvent(vLog)
+					if ev.OutputIndex.Sign() == 1 && c.isRelatedChallenge(ev.Asserter, ev.Challenger) {
+						c.log.Info("start to handle challenge", "outputIndex", ev.OutputIndex)
+						c.wg.Add(1)
+						go c.handleChallenge(ev.OutputIndex)
+					}
+				}
+			}
+			return
+		case <-c.ctx.Done():
+			return
 		}
 	}
+}
 
-	for i := c.checkpoint; i.Cmp(latestOutputIndex) != 1; i.Add(i, common.Big1) {
-		output, err := c.l2ooContract.GetL2Output(c.callOpts, i)
-		if err != nil {
-			return nil, err
-		}
+// subscribeL2OutputSubmitted subscribes the OutputSubmitted event from L2OutputOracle contract.
+// If the L2 output root is invalid, create challenge.
+// This function should be called only when challenger mode is on.
+func (c *Challenger) subscribeL2OutputSubmitted() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(time.Minute)
+	defer ticker.Stop()
+
+	for {
+	Loop:
+		select {
+		case <-ticker.C:
+			opts := &bind.WatchOpts{Context: c.ctx}
+			eventChan := make(chan *bindings.L2OutputOracleOutputSubmitted)
+			_, err := c.l2ooContract.WatchOutputSubmitted(opts, eventChan, nil, nil, nil)
+			if err != nil {
+				c.log.Error("failed to subscribe to output submitted event logs", "err", err)
+				break Loop
+			}
 
-		knownOutput, err := c.OutputAtBlockSafe(output.L2BlockNumber.Uint64(), false)
-		if err != nil {
-			return nil, err
+			for {
+				select {
+				case ev := <-eventChan:
+					c.log.Info("validating output", "l2BlockNumber", ev.L2BlockNumber, "outputRoot", ev.OutputRoot, "outputIndex", ev.L2OutputIndex)
+					// validate all outputs in between the checkpoint and the current outputIndex
+					for i := c.checkpoint; i.Cmp(ev.L2OutputIndex) != 1; new(big.Int).Add(i, common.Big1) {
+						c.wg.Add(1)
+						go c.handleOutput(i)
+					}
+					c.checkpoint = ev.L2OutputIndex
+				case <-c.ctx.Done():
+					return
+				}
+			}
```

Comment:
@kangsorang 
As you suggested, I made this simpler by using the subscribe function which is generated by bindings.
With this, we don't need to handle error from subscriber any more :)